### PR TITLE
Add tensorflow serving

### PR DESCRIPTION
**Pull request by hoanganhpham1006**

### CODE REVIEW

File: api/helpers/detect.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    return max(min(temp, 1), 0)
+
+def _interval_overlap(interval_a, interval_b):
+    x1, x2 = interval_a
+    x3, x4 = interval_b
+
+    if x3 < x1:
+        if x4 < x1:
+            return 0
+        else:
```
Comment (thandongtb) [inline api/helpers/detect.py]: có return thì không cần else nữa

File: api/helpers/get_embedding.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from fashion.embedding import Embedding
+
+request_timeout = settings.REQUEST_TIME_OUT_GET_EMB
+
+def client_request(host, port, in_tensor_name, in_tensor_dtype, input_data, model_spec_name):
+    img = input_data
+    img = img.reshape(-1, img.shape[1], img.shape[0], 3)
+    channel = implementations.insecure_channel(host, int(port))
+    stub = prediction_service_pb2.beta_create_PredictionService_stub(channel)
+    data = {'in_tensor_name': in_tensor_name, 'in_tensor_dtype': in_tensor_dtype, 'data': img}
```
Comment (thandongtb) [inline api/helpers/get_embedding.py]: sao không dùng thẳng luôn 

```
in_tensor_name, in_tensor_dtype, input_data,
```
mà phải truyền vào biến này

File: api/helpers/base64_validator.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import cv2
+
+def valid(img, error = 0):
```
Comment (thandongtb) [inline api/helpers/base64_validator.py]: hàm này em cho error = 0 trong thân hàm, không nên đưa vào tham số đầu vào của hàm nếu không xử lý đến

File: api/helpers/get_embedding.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    tensor_proto = tf.make_tensor_proto(img, dtype=in_tensor_dtype)
+    request.inputs[in_tensor_name].CopyFrom(tensor_proto)
+    predict_response = stub.Predict(request, timeout=request_timeout)
+    return predict_response
+
+def object_detect_img(img):
+    img_h = img.shape[0]
+    img_w = img.shape[1]
+    img_resize = cv2.resize(img, (settings.YOLO_IMG_SIZE, settings.YOLO_IMG_SIZE))
+    predict_response = client_request(host=settings.HOST_OBJ_DETECT, port=settings.PORT_OBJ_DETECT , in_tensor_name='images',
```
Comment (thandongtb) [inline api/helpers/get_embedding.py]: cho try catch vào timeout vào đề phòng trường hợp service tensorflow_serving chết

File: api/helpers/get_embedding.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            max_score = box['score']
+    if max_score > -1:
+        img = img[y_min:y_max, x_min:x_max]
+    return img
+
+def get_emb(img, object_detect):
+    emb = Embedding()
+    if object_detect:
+        img = object_detect_img(img)
+    predict_response = client_request(host=settings.HOST_GET_EMB, port=settings.PORT_GET_EMB, in_tensor_name='images', in_tensor_dtype=np.float32,
```
Comment (thandongtb) [inline api/helpers/get_embedding.py]: cũng cho try catch và timeout vào

File: api/views/similarity_detect.py [hunk 6]
```diff
@@ -30,22 +30,21 @@ def post(self, request):
...
-                embedding = output.decode("utf-8")
-                results = FashionDocument.search().query("match", embedding=embedding)
-                db.delete(k)
-                break
-            if (time.time() - start_time > 10):
-                return format(code=405, data=[], message=self.timeout, errors=[])
-
-            time.sleep(settings.CLIENT_SLEEP)
+        results = FashionDocument.search().query("match", embedding=embedding)
+        if (time.time() - start_time > 10):
```
Comment (thandongtb) [inline api/views/similarity_detect.py]: thằng này trước dùng trong vòng while để tính timeout giờ không cần nữa. Anh thấy nó sẽ không chạy vào case này bao giờ

File: api/helpers/detect.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+        return self.label
+
+    def get_score(self):
+        if self.score == -1:
+            self.score = self.classes[self.get_label()]
+
+        return self.score
+
+def _sigmoid(x):
```
Comment (thandongtb) [inline api/helpers/detect.py]: không dùng https://www.tensorflow.org/api_docs/python/tf/sigmoid được à ???. Softmax bên dưới tương tự
